File: uithelp/frontend/lib/data/models/locket_model.dart
class LocketModel {
  final String id;
  final String userId;
  final String userName;
  final String imageUrl;
  final String? caption;
  final int reactionsCount;
  final DateTime createdAt;

  const LocketModel({
    required this.id,
    required this.userId,
    required this.userName,
    required this.imageUrl,
    this.caption,
    required this.reactionsCount,
    required this.createdAt,
  });

  factory LocketModel.fromJson(Map<String, dynamic> json) => LocketModel(
        id: json['_id'] as String? ?? '',
        userId: json['userId'] as String? ?? '',
        userName: json['userName'] as String? ?? '',
        imageUrl: json['imageUrl'] as String? ?? '',
        caption: json['caption'] as String?,
        reactionsCount: (json['reactionsCount'] as num?)?.toInt() ?? 0,
        createdAt: json['createdAt'] != null
            ? DateTime.tryParse(json['createdAt']) ?? DateTime.now()
            : DateTime.now(),
      );
}

class LocketReactionModel {
  final String id;
  final String userId;
  final String userName;
  final String type;

  const LocketReactionModel({
    required this.id,
    required this.userId,
    required this.userName,
    required this.type,
  });

  factory LocketReactionModel.fromJson(Map<String, dynamic> json) =>
      LocketReactionModel(
        id: json['_id'] as String? ?? '',
        userId: json['userId'] as String? ?? '',
        userName: json['userName'] as String? ?? '',
        type: json['type'] as String? ?? 'like',
      );
}

class LocketCommentModel {
  final String id;
  final String userId;
  final String userName;
  final String content;
  final DateTime createdAt;

  const LocketCommentModel({
    required this.id,
    required this.userId,
    required this.userName,
    required this.content,
    required this.createdAt,
  });

  factory LocketCommentModel.fromJson(Map<String, dynamic> json) =>
      LocketCommentModel(
        id: json['_id'] as String? ?? '',
        userId: json['userId'] as String? ?? '',
        userName: json['userName'] as String? ?? '',
        content: json['content'] as String? ?? '',
        createdAt: json['createdAt'] != null
            ? DateTime.tryParse(json['createdAt']) ?? DateTime.now()
            : DateTime.now(),
      );
}

File: uithelp/backend/src/modules/locket/locket.model.ts
import mongoose from "mongoose";

const LocketSchema = new mongoose.Schema(
    {
        userId: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "User",
            required: true
        },
        userName: { type: String, default: "" },
        imageUrl: { type: String, required: true },
        caption: { type: String, default: null },
        reactionsCount: { type: Number, default: 0 },

        expiresAt: { type: Date, default: () => Date.now() + 7 * 24 * 60 * 60 * 1000 },
        isDeleted: { type: Boolean, default: false },
        deletedAt: { type: Date, default: null },
    },
    {
        timestamps: true
    }
);
const LocketReactionSchema = new mongoose.Schema(
    {
        locketId: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "Locket",
            required: true
        },
        userId: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "User",
            required: true
        },
        userName: { type: String, default: "" },
        type: {
            type: String,
            required: true,
            enum: ['like', 'heart', 'smile', 'sad'],
            default: 'like'
        },
        isDeleted: { type: Boolean, default: false },
        deletedAt: { type: Date, default: null },
    },
    {
        timestamps: true
    }
)

const LocketCommentSchema = new mongoose.Schema(
    {
        locketId: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "Locket",
            required: true
        },
        userId: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "User",
            required: true
        },
        userName: { type: String, default: "" },
        content: { type: String, required: true },
        commentsCount: { type: Number, default: 0 },
    },
    {
        timestamps: true
    }
)

LocketSchema.index({ expiresAt: 1 }, { expireAfterSeconds: 0 });
LocketReactionSchema.index(
    { locketId: 1, userId: 1 },
    { unique: true }
);

export const Locket = mongoose.model("Locket", LocketSchema);
export const LocketReaction = mongoose.model("LocketReaction", LocketReactionSchema);
export const LocketComment = mongoose.model("LocketComment", LocketCommentSchema);